### DISCUSSION TITLE
`dune build @check`, but only for a specific target

### DISCUSSION THREAD

**Discussion by ELLIOTTCABLE** [ocaml/dune#6969] [Q&A]
At the moment, there doesn't seem to be an idiomatic/beautiful way to build _just_ the `cmt{,i}`s for a _specific target_.

For example, if one usually builds `dune build src/api.bc` for a given subproject instead of the full `dune build` or `dune build src`, I can't figure out an equivalent to `dune build @src/api.bc/check` or what-have-you.

I can emulate the behaviour to some extent with something like `dune build src/.api.eobjs/byte/dune__exe__Api.cm{i,t,ti}`, but

1. there's no [variables](https://dune.readthedocs.io/en/stable/advanced-topics.html#variables-for-artifacts-1) for `.cmt{,i}`, so it's pretty ugly, fragile, and tied to Dune's implementation; and

2. and it doesn't seem like those can be generated for executable artifacts _anyway_, unless I'm mis-using them?

   ```
   opam exec --sw=../../ -- dune build '%{cmi:src/api.bc}'
   Entering directory '/home/me/code/monorepo'
   File "command line", line 1, characters 0-17:
   Error: Module Api.bc does not exist.
   make: *** [Makefile:39: check] Error 1
   ```

Is there a blessed path towards something like this? "Just typecheck the things that would be necessary to build <arbitrary executable target>?"

**Comment by Alizter** [comment]
What you can do is define a `(rule ...)` which has an `(alias foo)` that depends on the specific file in question. That way you can write `%{cmi:...}` in the deps. Afterwards you can just do `dune build @foo`.

**Comment by emillon** [comment]
Aliases are scoped by directories, so you can do `dune build @src/check` and will build only the `@check` targets for what's in `src/`.

**Reply by ELLIOTTCABLE** [reply]
Unfortunately, our gigantic monorepo often has lots of fiddly, precise build-targets that share directories. 😩

**Comment by rgrinberg** [comment]
I think this should work if you use a module name: `dune build '%{cmi:src/api.bc}'`

So for example: `dune build %{cmi:src/foo}`

Although note that this doesn't type check `Foo`. To do that, you would need to build `dune build %{cmo:src/foo}`
